Update localstack plugin to use LOCALSTACK_AUTH_TOKEN

simonrw opened the pull request:
## Overview
<!--  
Provide a high-level description of this change.   
-->

LocalStack is migrating to use auth tokens instead of api keys. This should be reflected in the 1Password LocalStack shell plugin.

See the first item in the [LocalStack v3.0.0 release notes miscellaneous section](https://blog.localstack.cloud/2023-11-16-announcing-localstack-30-general-availability/#miscellaneous).

Updates the plugin created in https://github.com/1Password/shell-plugins/pull/371


## Type of change
<!--  
Check the box below that describes your change best:
--> 

- [ ] Created a new plugin
- [x] Improved an existing plugin
- [ ] Fixed a bug in an existing plugin
- [ ] Improved contributor utilities or experience

## How To Test
<!--
Provide testing instructions for validating the changes introduced in this PR.
This will serve as a starting point for your reviewers, for functional testing.

If you created a new plugin, you can add a command here which can be used to test authentication.
For example, for the AWS CLI:
  aws s3 ls
-->

```
localstack start
```


## Changelog
<!--  
A one line sentence describing the change that this PR introduces. 
If this has impact over the user experience, your changelog will be included in the release notes of the next stable version of 1Password CLI.

Here are a few guidelines for writing a good changelog:
- Keep your description to a single sentence if you can, and use proper capitalization and punctuation, including a final period.
- Don't use emoji in your description.
- Avoid starting your sentence with "improved" or "fixed". Instead, describe the improvement or say what you fixed.
- Avoid using terminology like "Users are shown" or "You can now" and instead focus on the thing that was changed.

A few examples:

Authenticate the AWS CLI using Touch ID and other unlock options with 1Password Shell Plugins.
The AWS plugin can now be correctly initialized with a default credential, using `op plugin init`.
The AWS plugin now checks for the `AWS_SHARED_CREDENTIALS_FILE` environment variable and attempts to import credentials using the specified file.

For more examples, have a look over 1Password CLI's past release notes: 
https://app-updates.agilebits.com/product_history/CLI2
-->  

The LocalStack plugin now uses `LOCALSTACK_AUTH_TOKEN` instead of `LOCALSTACK_API_KEY`

simonrw committed:
Update localstack plugin to use LOCALSTACK_AUTH_TOKEN

scottisloud commented:
Thanks for updating the LocalStack plugin for auth tokens, and thank you for your patience. 

A few changes before merge:

**Env-var mapping**

`AuthToken()` and `APIKey()` currently share `defaultEnvVarMapping` with cross-field entries (`LOCALSTACK_API_KEY` → `fieldname.APIKey`, `LOCALSTACK_AUTH_TOKEN` → `fieldname.AuthToken`). Each credential type should have its own map — `AuthToken` should provision/import only `LOCALSTACK_AUTH_TOKEN`. Please also either remove the unused `APIKey()` path or update `api_key_test.go` if you keep it for backward compat.

**Auth gating**

Please add `needsauth.NotWhenContainsArgs("auth")` in `localstack.go` so setup commands like `localstack auth set-token`, `show-token`, and `clear-token` do not require a 1Password item first.

Optional: add `Prefix: "ls-"` to the token composition, and consider provisioning both `LOCALSTACK_AUTH_TOKEN` and `LOCALSTACK_API_KEY` during the migration window (LocalStack still accepts auth tokens via the legacy env var temporarily).

Happy to take another look once those are in!

simonrw commented:
Thanks @scottisloud , I am afk at the moment but I'll make these changes in a few weeks

simonrw committed:
Only support auth tokens

simonrw commented:
Thanks @scottisloud for the review comments. I have implemented the changes you suggested. We have stopped supporting API keys entirely (they are no longer given out by our authentication system), so I have removed the api key path altogether.

scottisloud commented:
Hi @simonrw 

I'll merge this shortly. Thanks for the patience here, and for the careful follow-up.

scottisloud reviewed (approved):
LGTM! 

Thanks again, and keep your eyes out for this to show up in a future release of the 1Password CLI!
